File: src/MagnusAddon/language_services/janome_ex/tokenizing/pre_processing_stage/word_info.py
from __future__ import annotations

from ankiutils import app
from autoslot import Slots
from language_services.jamdict_ex.dict_lookup import DictLookup
from language_services.janome_ex.tokenizing.pre_processing_stage.word_info_entry import WordInfoEntry


class WordInfo(Slots):
    @staticmethod
    def lookup(word: str) -> WordInfoEntry | None:
        vocab_entries = app.col().vocab.with_form(word)
        if vocab_entries.any():
            for vocab in vocab_entries:
                if vocab.get_question() == word:
                    return WordInfoEntry(word, vocab.parts_of_speech.get())
            return WordInfoEntry(word, vocab_entries.first().parts_of_speech.get())

        dict_lookup_result = DictLookup.lookup_word(word)
        if dict_lookup_result.found_words():
            return WordInfoEntry(word, dict_lookup_result.parts_of_speech())

        return None

    @staticmethod
    def lookup_godan(word: str) -> WordInfoEntry | None:
        word_info = WordInfo.lookup(word)
        return word_info if word_info is not None and word_info.is_godan else None

    @staticmethod
    def lookup_ichidan(word: str) -> WordInfoEntry | None:
        word_info = WordInfo.lookup(word)
        return word_info if word_info is not None and word_info.is_ichidan else None

    @staticmethod
    def is_godan(word: str) -> bool: return WordInfo.lookup_godan(word) is not None

    @staticmethod
    def is_ichidan(word: str) -> bool: return WordInfo.lookup_ichidan(word) is not None

File: src/src_dotnet/JAStudio.Core/LanguageServices/JanomeEx/Tokenizing/PreProcessingStage/WordInfo.cs
using System.Linq;
using JAStudio.Core.LanguageServices.JamdictEx;
using JAStudio.Core.Note.Collection;

namespace JAStudio.Core.LanguageServices.JanomeEx.Tokenizing.PreProcessingStage;

public static class WordInfo
{
   public static WordInfoEntry? Lookup(VocabCollection vocab, DictLookup dictLookup, string word)
   {
      var vocabEntries = vocab.WithForm(word);
      if(vocabEntries.Any())
      {
         // Try to find exact question match first
         foreach(var v in vocabEntries)
         {
            if(v.GetQuestion() == word)
            {
               return new VocabWordInfoEntry(word, v);
            }
         }

         return new VocabWordInfoEntry(word, vocabEntries.First());
      }

      var dictLookupResult = dictLookup.LookupWord(word);
      if(dictLookupResult.FoundWords())
      {
         return new DictWordInfoEntry(word, dictLookupResult);
      }

      return null;
   }

   public static WordInfoEntry? LookupGodan(VocabCollection vocab, DictLookup dictLookup, string word)
   {
      var wordInfo = Lookup(vocab, dictLookup, word);
      return wordInfo != null && wordInfo.IsGodan ? wordInfo : null;
   }

   public static WordInfoEntry? LookupIchidan(VocabCollection vocab, DictLookup dictLookup, string word)
   {
      var wordInfo = Lookup(vocab, dictLookup, word);
      return wordInfo != null && wordInfo.IsIchidan ? wordInfo : null;
   }

   public static bool IsGodan(VocabCollection vocab, DictLookup dictLookup, string word) => LookupGodan(vocab, dictLookup, word) != null;

   public static bool IsIchidan(VocabCollection vocab, DictLookup dictLookup, string word) => LookupIchidan(vocab, dictLookup, word) != null;
}
